api/firstbook2.py

```diff
@@ -1,31 +1,67 @@
-import logging
+from __future__ import annotations
+
+import re
 import time
-import urllib.parse
-from typing import Optional, Union
+from typing import Optional, Pattern, Union
 
 import jwt
 import requests
 from flask_babel import lazy_gettext as _
+from pydantic import HttpUrl
 
-from core.model import ExternalIntegration, Patron
+from core.integration.settings import ConfigurationFormItem
+from core.model import Patron
 
-from .authenticator import BasicAuthenticationProvider, PatronData
+from .authenticator import (
+    BasicAuthenticationProvider,
+    BasicAuthenticationProviderLibrarySettings,
+    BasicAuthenticationProviderSettings,
+    PatronData,
+)
 from .circulation_exceptions import RemoteInitiatedServerError
-from .config import CannotLoadConfiguration
+
+
+class FirstBookAuthenticationSettings(BasicAuthenticationProviderSettings):
+    class ConfigurationForm(BasicAuthenticationProviderSettings.ConfigurationForm):
+        url = ConfigurationFormItem(
+            label=_("URL"),
+            description=_("The URL for the First Book authentication service."),
+            required=True,
+        )
+
+        password = ConfigurationFormItem(
+            label=_("Key"),
+            description=_("The key for the First Book authentication service."),
+        )
+
+    url: HttpUrl = "https://ebooksprod.firstbook.org/api/"  # type: ignore[assignment]
```

Comment:
Do we need to define these fields twice? Is there a way we can do away with ConfigurationForm? and have the BaseSettings object hold the attributes?

Reply:
I'm open to suggestions on other ways to approach this. I chose the approach of putting an additional `ConfigurationForm` object on the Pydantic class since it aligns with how Pydantic does its configuration. It also separates the form information from the dataclass a bit, which might be useful if we want some configuration fields that are not part of the form. However, I also don't like that we have to define these fields twice. 

Another approach that I thought about was to override the Pydantic `Field` class, so that the class definition could look something like this:

```python
class FirstBookAuthenticationSettings(BasicAuthenticationProviderSettings):
    url: HttpUrl = Field(
        "https://ebooksprod.firstbook.org/api/",
        form_item=ConfigurationFormItem(
            label=_("URL"),
            description=_("The URL for the First Book authentication service."),
            required=True,
        )
    )
```

The problem I ran into with this approach is that `Field` is actually a function, that does some validation and then initializes a class. So we would need to override a couple core Pydantic functions / classes to make this work. I was a bit uncomfortable getting that much into the guts of Pydantic here, but it does provide cleaner definitions in our code with less repetition. 

Let me know what you think, or if you have other approaches you suggest. I don't mind revisiting how we define these settings classes.

Reply:
The pydantic [schema customization](https://docs.pydantic.dev/latest/usage/schema/#field-customization) document has the following for `Field()`
```
** any other keyword arguments (e.g. examples) will be added verbatim to the field's schema
```

So this code works
```
class SomeSettings:
    value: str = Field(config_settings="any-additional-settings", description="A pydantic description")

print (SomeSettings.value.extra)
```
This outputs `{'config_settings': 'any-additional-settings'}`

Note: The `**` is now deprectated in favour of a special keyword `json_schema_extra` [source](https://github.com/pydantic/pydantic/blob/main/pydantic/fields.py#L607)

I propose defining a method as follows:
```
def ConfigField(settings=None, ... all the Field kwargs...):
    return Field(... all the kwargs..., config_settings=settings)
```
We can easily change this to `json_schema_extra=dict(config_settings=settings)` once we move to Pydantic V2 (which is still in alpha).